packages/react-data-grid/src/Canvas.tsx

```diff
@@ -168,59 +151,34 @@ export default function Canvas<R>({
   }
 
   function getRows() {
-    const rows = [];
+    const rowElements = [];
 
     for (let idx = rowOverscanStartIdx; idx <= rowOverscanEndIdx; idx++) {
-      rows.push(renderRow(idx));
-    }
-
-    return rows;
-  }
-
-  function renderRow(idx: number) {
-    const row = rowGetter(idx);
-    const rendererProps: RendererProps<R> = {
-      key: idx,
-      ref(row) {
-        if (row) {
-          rows.set(idx, row);
-        } else {
-          rows.delete(idx);
-        }
-      },
-      idx,
-      row,
-      height: rowHeight,
-      columns: columnMetrics.columns,
-      isRowSelected: isRowSelected(row),
-      onRowSelectionChange,
-      cellMetaData,
-      subRowDetails: getSubRowDetails ? getSubRowDetails(row) : undefined,
-      colOverscanStartIdx,
-      colOverscanEndIdx,
-      lastFrozenColumnIndex: columnMetrics.lastFrozenColumnIndex,
-      scrollLeft
-    };
-    const { __metaData } = row as RowData;
-
-    if (__metaData) {
-      if (__metaData.getRowRenderer) {
-        return __metaData.getRowRenderer(rendererProps, idx);
-      }
-      if (__metaData.isGroup) {
-        return renderGroupRow(rendererProps);
-      }
-    }
-
-    if (rowRenderer) {
-      return renderCustomRowRenderer(rendererProps);
+      const rowData = rowGetter(idx);
+      rowElements.push(
+        <RowRenderer<R>
+          key={idx}
+          idx={idx}
+          rowData={rowData}
+          rowRefs={rowRefs}
```

Comment:
Instead of passing the `rowRefs`, should we pass a ref setting function? It seems weird for each individual row to have access to all other row refs.

Reply:
If I did this we'd pass a new ref function on each render, which might trip up `memo`, and i'm having trouble typing the ref parameter for some reason.

Reply:
In `Canvas.tsx`
```typescript
const setRowRef = useCallback((rowRef: (IRowRenderer & React.Component<IRowRendererProps<R>>) | null, idx: number) => {
    if (rowRef) {
      rowRefs.set(idx, rowRef);
    } else {
      rowRefs.delete(idx);
    }
  }, [rowRefs]);

<RowRenderer<R>
  key={idx}
  idx={idx}
  rowData={rowData}
  setRowRef={setRowRef}
  ...
/>
```

In `RowRenderer.tsx`
```typescript
interface RowRendererProps<R> extends SharedCanvasProps<R> {
  ...
  setRowRef(rowRef: (IRowRenderer & React.Component<IRowRendererProps<R>>) | null, idx: number): void;
}

const rendererProps: RendererProps<R> = {
    ref(row) {
      setRowRef(row, idx);
    },
   ...
};
```

This should work isn't it?